src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs

```diff
@@ -454,33 +289,93 @@ internal bool IsExpressionBodied
                 {
-                    var diagnostics = DiagnosticBag.GetInstance();
-                    var result = this.ComputeType(binder: null, CSharpSyntaxNode, diagnostics);
-                    if (Interlocked.CompareExchange(ref _lazyType, new TypeWithAnnotations.Boxed(result), null) == null)
+                    if (_state.NotePartComplete(CompletionPart.StartPropertyEnsureSignature))
+                    {
+                        // By setting StartPropertyEnsureSignature, we've committed to doing the work and setting
+                        // FinishPropertyEnsureSignature.  So there is no cancellation supported between one and the other.
+                        var diagnostics = DiagnosticBag.GetInstance();
+                        try
+                        {
+                            EnsureSignature(diagnostics);
+                            AddDiagnostics(diagnostics);
+                        }
+                        finally
+                        {
+                            _state.NotePartComplete(CompletionPart.FinishPropertyEnsureSignature);
+                            diagnostics.Free();
+                        }
+                    }
+                    else
                     {
-                        this.AddDeclarationDiagnostics(diagnostics);
+                        // Either (1) this thread is in the process of completing the method,
+                        // or (2) some other thread has beat us to the punch and completed the method.
+                        // We can distinguish the two cases here by checking for the FinishPropertyEnsureSignature
+                        // part to be complete, which would only occur if another thread completed this
+                        // method.
+                        //
+                        // The other case, in which this thread is in the process of completing the method,
+                        // requires that we return here even though the work is not complete. That's because
+                        // signature is processed by first populating the return type and parameters by binding
+                        // the syntax from source.  Those values are visible to the same thread for the purpose
+                        // of computing which methods are implemented and overridden.  But then those values
+                        // may be rewritten (by the same thread) to copy down custom modifiers. In order to
+                        // allow the same thread to see the return type and parameters from the syntax (though
+                        // they do not yet take on their final values), we return here.
+
+                        // Due to the fact that this method is potentially reentrant, we must use a 
+                        // reentrant lock to avoid deadlock and cannot assert that at this point the work
+                        // has completed (_state.HasComplete(CompletionPart.FinishPropertyEnsureSignature)).
```

Comment:
> How do we guard against this type of problem? 

All the fields that are expected to be used in case of reentrancy are set by EnsureSignature to some preliminary values before it does anything that could cause reentrancy. We are already using similar mechanism for method symbols.

---
In reply to: [525320776](https://github.com/dotnet/roslyn/pull/49360#discussion_r525320776) [](ancestors = 525320776)